Added section on common unsafe operations and their safe counterparts (#8654)
godotengine/godot-docs: tutorials/scripting/gdscript/static_typing.rst

``UNSAFE_*`` warnings make unsafe operations more noticeable, than unsafe lines.
Currently, ``UNSAFE_*`` warnings do not cover all cases that unsafe lines cover.

Common unsafe operations and their safe counterparts
----------------------------------------------------

The example below gives an ``UNSAFE_CAST`` warning because ``body.health`` is of

> an unknown type:

Reviewer: ```suggestion
an unknown type::
```
This is required to make it a block, see below and above

    func _on_body_entered(body: Node2D) -> void:
        if "health" in body:
            (body.health as Health).take_damage(damage)

This can be made safe using the following syntax, which works because 
``Object.get()`` returns ``null`` if the property doesn't exist or is of another

> type:

Reviewer: ```suggestion
type::
```

    func _on_body_entered(body: Node2D) -> void:
	    var health: Health = body.get("health")
	    if health != null:
	    	health.take_damage(damage)

In another example below, we call a method that we know is on the object or
script, but we will still get an ``UNSAFE_METHOD_ACCESS`` warning as the method

> is not present in the inferred type (in this case being a Node2D):

Reviewer: ```suggestion
is not present in the inferred type (in this case being a Node2D)::
```

>     if node2d.has_method("some_function"):
>         node2d.some_function()

Reviewer: ```suggestion
    if node_2d.has_method("some_function"):
        node_2d.some_function()
```

> To make this operation safe, you can use the following syntax:

Reviewer: ```suggestion
To make this operation safe, you can use the following syntax::
```

>     var myScript: MyScript = node2d
>     if myScript != null:
>         myScript.some_function()

Reviewer: ```suggestion
    var my_script: MyScript = node_2d
    if my_script != null:
        my_script.some_function()
```
Use `snake_case` for GDScript

Reviewer: It's easy with the PascalCase named types 🙂

``UNSAFE_*`` warnings make unsafe operations more noticeable, than unsafe lines.
Currently, ``UNSAFE_*`` warnings do not cover all cases that unsafe lines cover.

Common unsafe operations and their safe counterparts
----------------------------------------------------

Warning: ``UNSAFE_CAST``
~~~~~~~~~~~~~~~~~~~~~~~~

In the example below, we would like an object entering our collision area to take

> damage if it has a health variable that contains a ``Health`` object. The physics

Reviewer 2: I think most often the `health` variable is of type `int` or `float`. Perhaps this example is not very successful in a practical sense. I often cite the `_input()` method and the `InputEvent` and `InputEventMouseMotion` classes as an example of appropriate use of the `as` operator.

Author: Good point; came up with another scenario.

system passes us a ``Node2D`` object, and after we check whether that object has
the health variable, we cast the variable to a ``Health`` object and call
``take_damage()``. However, this gives an ``UNSAFE_CAST`` warning because

> ``body.health`` is of an unknown type::

Reviewer 2: ```suggestion
``body.health`` is of ``Variant`` type::
```

    func _on_body_entered(body: Node2D) -> void:
        if "health" in body:
            (body.health as Health).take_damage(damage)

This operation can be made type safe using the ``Object.get()`` method.

> ``body.get("health")`` will return ``null`` if the property doesn't exist or if
> it is of another type::

Reviewer 2: ```suggestion
``body.get("health")`` will silently return ``null`` if the property doesn't exist::
```

When calling `get()`, you cannot specify what type you are expecting. This is not a special syntax, but a regular method ([`Object.get()`](https://docs.godotengine.org/en/stable/classes/class_object.html#class-object-method-get)).

    func _on_body_entered(body: Node2D) -> void:
        var health: Health = body.get("health")
        if health != null:

>             health.take_damage(damage)

Reviewer 2: It is not safe. The safe option is

```gdscript
        var health := body.get("health") as Health
        if health != null:
            health.take_damage(damage)
```

or

```gdscript
        if body.get("health") is Health:
            var health: Health = body.get("health")
            health.take_damage(damage)
```

Note that `body.get("health")` is executed twice, so a more correct option is

```gdscript
        var health_variant: Variant = body.get("health")
        if health_variant is Health:
            var health: Health = health_variant
            health.take_damage(damage)
```

Warning: ``UNSAFE_METHOD_ACCESS``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In the example below, we aim to call a method named ``some_function()`` on a
script with ``class_name MyScript`` and that ``extends Node2D``. If we have a 
reference to the object as a ``Node2D`` (for instance, as it was passed to us by
the physics system), we can first check if the method exists and then call the
method if it does. However, we will get an ``UNSAFE_METHOD_ACCESS`` warning 
as the method is not present in the referenced type (in this case a ``Node2D``)::

    if node_2d.has_method("some_function"):
        node_2d.some_function()

To make this operation safe, you can first check if the object is of type

> ``MyScript``and then declare a variable with the type ``MyScript`` on which you

Reviewer 2: ```suggestion
``MyScript`` and then declare a variable with the type ``MyScript`` on which you
```

``UNSAFE_*`` warnings make unsafe operations more noticeable, than unsafe lines.
Currently, ``UNSAFE_*`` warnings do not cover all cases that unsafe lines cover.

Common unsafe operations and their safe counterparts
----------------------------------------------------

> Warnings: ``UNSAFE_PROPERTY_ACCESS`` and ``UNSAFE_METHOD_ACCESS``
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer 2: ```suggestion
``UNSAFE_PROPERTY_ACCESS`` and ``UNSAFE_METHOD_ACCESS`` warnings
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Just a suggestion. I'm not sure what's best, but the colon in the header looks a bit weird to me.

In this example, we aim to set a property and call a method on an object
that has a script attached with ``class_name MyScript`` and that ``extends
Node2D``. If we have a reference to the object as a ``Node2D`` (for instance,
as it was passed to us by the physics system), we can first check if the
property and method exist and then set and call them if they do::
    

>     if "some_property" in node_2d:
>         node_2d.some_property = 20  # gives UNSAFE_PROPERTY_ACCESS warning
>     
>     if node_2d.has_method("some_function"):
>         node_2d.some_function()  # gives UNSAFE_METHOD_ACCESS warning

Reviewer 2: ```suggestion
    if "some_property" in node_2d:
        node_2d.some_property = 20 # Produces UNSAFE_PROPERTY_ACCESS warning.
    
    if node_2d.has_method("some_function"):
        node_2d.some_function() # Produces UNSAFE_METHOD_ACCESS warning.
```

> However, this code will return ``UNSAFE_PROPERTY_ACCESS`` and

Reviewer 2: ```suggestion
However, this code will produce ``UNSAFE_PROPERTY_ACCESS`` and
```

I'm not sure whether "gives" or "produces" is better, but "returns" looks incorrect to me.

``UNSAFE_METHOD_ACCESS`` warnings as the property and method are not present
in the referenced type - in this case a ``Node2D``. To make these operations
safe, you can first check if the object is of type ``MyScript`` using the
``is`` keyword and then declare a variable with the type ``MyScript`` on
which you can set its properties and call its methods::
    
    if node_2d is MyScript:
        var my_script: MyScript = node_2d
        my_script.some_property = 20
        my_script.some_function()

> 

Reviewer 2: ```suggestion

Alternatively, you can use the ``as`` operator::

    var my_script := node_2d as MyScript
    if my_script:
        my_script.some_property = 20
        my_script.some_function()
```

Author: I'll included this alternative in a new commit, but I do want to note that this produces a INFERRED_DECLARATION warning. It doesn't feel entirely right to include example solutions which produce warnings ;)

Reviewer 2: Fair point, but note that `INFERRED_DECLARATION` is added for those who prefer verbose code. This is a strict mode to the second power, we assume that few people enable the warning.

There is also currently no way to specify a metatype: for example if you preload a script or enum into a constant (use `const` as an analogue of `typedef`/`using` in C++). In this case, you are forced to trigger the `INFERRED_DECLARATION` warning and suppress it with `@warning_ignore`.

> Warning: ``UNSAFE_CAST``
> ~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer 2: ```suggestion
``UNSAFE_CAST`` warning
~~~~~~~~~~~~~~~~~~~~~~~
```

In this example, we would like the label connected to an object entering our
collision area to show the area's name. Once the object enters the collision
area, the physics system sends a signal with a ``Node2D`` object, and the most
straightforward (but not statically typed) solution to do what we want could
be achieved like this::
    
    func _on_body_entered(body: Node2D) -> void:

>         body.label.text = name  # gives UNSAFE_PROPERTY_ACCESS warning

Reviewer 2: ```suggestion
        body.label.text = name # Produces UNSAFE_PROPERTY_ACCESS warning.
```

This piece of code gives a ``UNSAFE_PROPERTY_ACCESS`` warning because ``label``
is not defined in ``Node2D``. To solve this, we could first check if the
``label`` property exist and cast it to type ``Label`` before settings its text
property like so::

    func _on_body_entered(body: Node2D) -> void:
        if "label" in body:

>             (body.label as Label).text = name  # gives UNSAFE_CAST warning

Reviewer 2: ```suggestion
            (body.label as Label).text = name # Produces UNSAFE_CAST warning.
```

However, this gives an ``UNSAFE_CAST`` warning because ``body.label`` is of a

> ``Variant`` type. This operation can be made type safe using the
> ``Object.get()`` method that returns the property as a ``Variant`` (or
> ``null`` if the property doesn't exist), using the ``is`` keyword to determine
> whether the property contains an object of the right type, and using a
> variable declaration with a static type::

Reviewer 2: ```suggestion
``Variant`` type. To safely get a property you can use the ``Object.get()``
method, which silently returns ``null`` if the property doesn't exist.
Then, since the statically method returns a ``Variant`` value, you can check
to determine whether the property contains an object of the right type, and
using a variable declaration with a static type::
```

The original wording contrasts `Variant` and `null`, although `null` is a `Variant` value too. It also seems like `Object.get()` can choose which static type it returns, but the static type is predefined and does not depend on whether the property exists (which is checked at runtime, not at compile time). I hope my suggestion removes potential confusion.

Author: Took your suggestion and slightly iterated on it; mostly added `using the ``is`` keyword` as it made the list of steps less vague.

that has a script attached with ``class_name MyScript`` and that ``extends
Node2D``. If we have a reference to the object as a ``Node2D`` (for instance,
as it was passed to us by the physics system), we can first check if the
property and method exist and then set and call them if they do::
    
    if "some_property" in node_2d:
        node_2d.some_property = 20  # Produces UNSAFE_PROPERTY_ACCESS warning.
    
    if node_2d.has_method("some_function"):
        node_2d.some_function()  # Produces UNSAFE_METHOD_ACCESS warning.

However, this code will produce ``UNSAFE_PROPERTY_ACCESS`` and
``UNSAFE_METHOD_ACCESS`` warnings as the property and method are not present
in the referenced type - in this case a ``Node2D``. To make these operations
safe, you can first check if the object is of type ``MyScript`` using the
``is`` keyword and then declare a variable with the type ``MyScript`` on
which you can set its properties and call its methods::
    
    if node_2d is MyScript:
        var my_script: MyScript = node_2d
        my_script.some_property = 20
        my_script.some_function()

Alternatively, you can use the ``as`` operator::

    var my_script := node_2d as MyScript

>     if my_script:

Author: @dalexeev sorry, last question! In my opinion, `if my_script:` reads as if my_script is a boolean. I know that it actually means `if my_script != null:` so I was thinking to write that out like that in this example. Agreed?

Reviewer 2: It's a bit complicated since in addition to `null` there is `Object#null` and `Freed Object` (see also godotengine/godot#73896), but I think that comparison with `null` will work here. There is also `is_instance_valid()`. On the other hand, implicit casting to `bool` has the advantage of fewer opcodes.

Author: Hmm, complicated indeed. Perhaps the most logical way of thinking about this particular case is that `my_script` is explicitly set to `null` if `node_2d` is not of type `MyScript`, and the explicitly verbose `null` check would help illustrate what happens when the cast fails. So even though I agree the fewer opcodes makes sense in most situations, perhaps in this example the slightly more verbose comparison to `null` makes sense in this example? I've pushed a new commit.
